File: CLR_UDF_CS/SQLFUN.cs
using System;
using System.Data;
using System.Data.SqlClient;
using System.Data.SqlTypes;
//using System.Transactions;
using Microsoft.SqlServer.Server;
using System.Collections;
using System.Collections.Generic;
using System.Text.RegularExpressions;
namespace CSUDF_SQLFUN {
    public static class funUtil {
        public static void ex(string connStr, string cmdStr, string db, string server, ref DataSet ds, int r, int cc, ref object dtout)
        {
            string catalog = db.ToString() == "" ? "master" : db;
            string c;
            SqlConnection connection;
            SqlCommand sqlcmd;
            if (connStr != null && Regex.IsMatch(connStr, "context\\s+connection=true"))
            {
                c = "context connection=true"; // + ";Initial Catalog=" + catalog + ";";
                connection = new SqlConnection(c);
                connection.Open();
                sqlcmd = new SqlCommand(cmdStr, connection);
                var rdr = sqlcmd.ExecuteReader();
                //SqlContext.Pipe.Send(rdr);

                //throw new Exception(rdr.Read().ToString());

                for (int ridx = 0; ridx <= r; ridx++){ rdr.Read(); };
                dtout = rdr.GetSqlValue(cc);
                rdr.Close();
                connection.Close();
                
            }
            else {
                string serverhost = server.ToString() == "" ? "localhost" : server;
                c = connStr == "" || connStr == null ? "Data Source=" + serverhost + "; Initial Catalog=" + catalog + "; Persist Security Info=True;Integrated Security=SSPI;Enlist=false;MultipleActiveResultSets=false;" : connStr;
                connection = new SqlConnection(c);
                connection.Open();
                sqlcmd = new SqlCommand(cmdStr, connection);
                var da = new SqlDataAdapter(sqlcmd);
                ds = new DataSet();
                var fill = da.Fill(ds);
                da.Dispose();
            }
            sqlcmd.Dispose();
            connection.Close();
            connection.Dispose();
        }

        public static void ex(string connStr, string cmdStr, string db, string server) {
            var ds = new DataSet();
            object o = null;
            ex(connStr, cmdStr, db, server, ref ds, -1, -1, ref o);
        }
        public static void ex(string connStr, string cmdStr, string db, string server, ref DataSet ds)
        {
            object o = null;
            ex(connStr, cmdStr, db, server, ref ds, -1, -1, ref o);
        }

        //public static DataSet ex1(string connStr, string cmdStr, string db, string server) {
        //    string c = "";
        //    SqlConnection connection = null;
        //    SqlCommand sqlcmd = null;
        //    string catalog = db.ToString() == "" ? "master" : db;
        //    if (connStr == "" || connStr == null) {
                
        //        string serverhost = server.ToString() == "" ? "." : server;
        //        c = "Data Source=" + serverhost + ";Initial Catalog=" + catalog + ";Persist Security Info=True;Integrated Security=SSPI;Enlist=false;MultipleActiveResultSets=false;";
        //        //c = connStr == "" ? "Data Source=localhost\\SQL16;Initial Catalog=master;Persist Security Info=True;Integrated Security=SSPI;Enlist=false;MultipleActiveResultSets=false;" : connStr;
        //        connection = new SqlConnection(c);
        //        connection.Open();
        //        sqlcmd = new SqlCommand(cmdStr, connection);
        //        var da = new SqlDataAdapter(sqlcmd);
        //        var ds = new DataSet();
        //        var fill = da.Fill(ds);
        //        da.Dispose();
        //        sqlcmd.Dispose();
        //        connection.Close();
        //        connection.Dispose();
        //        return ds;
        //    } else {
        //        //try
        //        //{
        //        //    c = connStr;
        //        //    connection = new SqlConnection(c);
        //        //    connection.Open();
        //        //}
        //        //catch
        //        //{
        //            c = "context connection=true;"; // Initial Catalog=" + catalog;
        //            connection = new SqlConnection(c);
        //            connection.Open();
        //        //}
                
        //        sqlcmd = new SqlCommand(cmdStr, connection);
        //        if (sqlcmd == null)
        //        {
        //            throw new Exception("1");
        //        }
        //        SqlDataReader rdr = sqlcmd.ExecuteReader();
        //        if (rdr == null)
        //        {
        //            throw new Exception("123");
        //        }
        //        try {
        //            SqlContext.Pipe.Send(rdr);
        //            //rdr.Close();
        //            //connection.Close();
        //        } catch
        //        {

        //        }
        //        var ds = new DataSet();
        //        return ds;
        //    }
            

            
            
        //}
    }
    public class UDF_SQLFUN {
        [SqlFunction(IsDeterministic = false, IsPrecise = true, DataAccess = DataAccessKind.Read)]
        public static SqlString SQLFUN(string conn, string cmd, int r, int c, string db, string server) {
            var ds = new DataSet();
            object o = null;
            funUtil.ex(conn, cmd, db, server, ref ds, r, c, ref o);
            if (conn == "" || conn == null)
            {
                var results = new ArrayList();
                var i = ds.Tables[0].Rows[r][c];
                if (i is System.DBNull)
                {
                    return System.Data.SqlTypes.SqlString.Null;
                }
                else {
                    return new System.Data.SqlTypes.SqlString(i.ToString());
                }
            } else
            {
                return (System.Data.SqlTypes.SqlString)o.ToString();
            }
        }

        [SqlFunction(IsDeterministic = false, IsPrecise = true)] 
        public static System.Nullable<int> INTFUN(string conn, string cmd, int r, int c, string db, string server) {
            var results = new ArrayList();
            var ds = new DataSet();
            funUtil.ex(conn, cmd, db, server, ref ds);
            var i = ds.Tables[0].Rows[r][c];
            if (i is System.DBNull) { 
                return new System.Nullable<int>();
            } else {
                return (System.Nullable<int>)i;
            }
        }

        [SqlFunction(IsDeterministic = false, IsPrecise = true)]
        public static SqlDouble FLOATFUN(string conn, string cmd, int r, int c, string db, string server) {
            var results = new ArrayList();
            var ds = new DataSet();
            funUtil.ex(conn, cmd, db, server, ref ds);
            var i = ds.Tables[0].Rows[r][c];
            if (i is System.DBNull) {
                return System.Data.SqlTypes.SqlDouble.Null;
            } else {
                return new System.Data.SqlTypes.SqlDouble((float)i);
            }
        }
        [SqlFunction(IsDeterministic = false, IsPrecise = true)]
        public static SqlSingle REALFUN(string conn, string cmd, int r, int c, string db, string server) {
            var results = new ArrayList();
            var ds = new DataSet();
            funUtil.ex(conn, cmd, db, server, ref ds);
            var i = ds.Tables[0].Rows[r][c];
            if (i is System.DBNull) {
                return System.Data.SqlTypes.SqlSingle.Null;
            } else {
                return new System.Data.SqlTypes.SqlSingle((Single)i);
            }
        }

        [SqlFunction(IsDeterministic = false, IsPrecise = true)]
        public static SqlDateTime DTFUN(string conn, string cmd, int r, int c, string db, string server) {
            var results = new ArrayList();
            var ds = new DataSet();
            funUtil.ex(conn, cmd, db, server, ref ds);
            var i = ds.Tables[0].Rows[r][c];
            if (i is System.DBNull) {
                return System.Data.SqlTypes.SqlDateTime.Null;
            } else {
                return new System.Data.SqlTypes.SqlDateTime((DateTime)i);
            }
        }

        [SqlFunction(IsDeterministic = false, IsPrecise = true)]
        public static SqlBoolean BITFUN(string conn, string cmd, int r, int c, string db, string server) {
            var results = new ArrayList();
            var ds = new DataSet();
            funUtil.ex(conn, cmd, db, server, ref ds);
            var i = ds.Tables[0].Rows[r][c];
            if (i is System.DBNull) {
                return System.Data.SqlTypes.SqlBoolean.Null;
            } else {
                return new System.Data.SqlTypes.SqlBoolean((bool)i);
            }
        }

    }
}

File: CLR_UDF/SQLFUN.fs
namespace FSUDF_SQLFUN
(*
#r "C:\\Program Files (x86)\\Reference Assemblies\\Microsoft\\Framework\\.NETFramework\\v4.6\\System.Transactions.dll"
#r "C:\\Program Files (x86)\\Reference Assemblies\\Microsoft\\Framework\\.NETFramework\\v4.6\\System.Data.dll"


*)
open System
open System.Data
open System.Data.SqlClient
open System.Data.SqlTypes
open System.Transactions
open Microsoft.SqlServer.Server
open System.Collections
open System.Collections.Generic

module funUtil =   
    let ex = fun connStr cmdStr ->

        match connStr with
        | "context connection=true" | _ when connStr.GetType() <> typeof<string> ->
            let c = "context connection=true"
            //let reader = sqlcmd.ExecuteReader()
            //(connection, sqlcmd, reader)
            let connection = new SqlConnection(c)
            connection.Open()
            let sqlcmd = new SqlCommand(cmdStr, connection)
            let rdr = sqlcmd.ExecuteReader()
            SqlContext.Pipe.Send(rdr) 
            rdr.Close()
            connection.Close()
            let ds = new DataSet()
            ds
        | _ (*when connStr.GetType() = typeof<string>*) ->
            let c = 
                match connStr with
                | "" -> 
                    "Data Source=localhost\SQL16;Initial Catalog=master;Persist Security Info=True;Integrated Security=SSPI;Enlist=false;MultipleActiveResultSets=false;"
                | _ ->
                    connStr
            let connection = new SqlConnection(c)
            connection.Open()
            let sqlcmd = new SqlCommand(cmdStr, connection)
            let da = new SqlDataAdapter(sqlcmd)
            let ds = new DataSet()
            let fill = da.Fill(ds)
            da.Dispose()
            sqlcmd.Dispose()
            connection.Close()
            connection.Dispose()
            ds
        ;

        //DataAccessKind.Read or SystemDataAccessKind.Read
type public UDF_SQLFUN = 
    [<SqlFunction(IsDeterministic = false, IsPrecise = true, DataAccess = DataAccessKind.Read)>]
        static member SQLFUN(conn : string, cmd : string, r : int, c : int) =
            let results = new ArrayList()
            match conn with
            | "context connection=true" | _ when conn.GetType() <> typeof<string> ->
                let last = funUtil.ex conn cmd
                System.Data.SqlTypes.SqlString.Null
            | _ ->
                let ds = funUtil.ex conn cmd
                let i = ds.Tables.[0].Rows.[r].Item(c) 
                match i with
                | :? System.DBNull ->
                    System.Data.SqlTypes.SqlString.Null
                | _ ->
                    //i :?> System.Data.SqlTypes.SqlString
                    new System.Data.SqlTypes.SqlString(i.ToString())


    [<SqlFunction(IsDeterministic = false, IsPrecise = true)>]
        static member INTFUN(conn : string, cmd : string, r : int, c : int) =
            let results = new ArrayList()
            let ds = funUtil.ex conn cmd
            let i = ds.Tables.[0].Rows.[r].Item(c) 
            match i with 
            //| _ when (i :?> System.DBNull) = System.DBNull.Value ->
            | :? System.DBNull ->
                new System.Nullable<int>()
            | _ ->
                i :?> System.Nullable<int>


    [<SqlFunction(IsDeterministic = false, IsPrecise = true)>]
        static member FLOATFUN(conn : string, cmd : string, r : int, c : int) =
            let results = new ArrayList()
            let ds = funUtil.ex conn cmd
            let i = ds.Tables.[0].Rows.[r].Item(c)
            match i with
            | :? System.DBNull ->
                System.Data.SqlTypes.SqlDouble.Null
            | _ ->
                new System.Data.SqlTypes.SqlDouble(i :?> float)

    [<SqlFunction(IsDeterministic = false, IsPrecise = true)>]
        static member REALFUN(conn : string, cmd : string, r : int, c : int) =
            let results = new ArrayList()
            let ds = funUtil.ex conn cmd
            let i = ds.Tables.[0].Rows.[r].Item(c)
            match i with
            | :? System.DBNull ->
                System.Data.SqlTypes.SqlSingle.Null
            | _ ->
                new System.Data.SqlTypes.SqlSingle(i :?> single)

    [<SqlFunction(IsDeterministic = false, IsPrecise = true)>]
        static member DTFUN(conn : string, cmd : string, r : int, c : int) =
            let results = new ArrayList()
            let ds = funUtil.ex conn cmd
            let i = ds.Tables.[0].Rows.[r].Item(c)
            match i with
            | :? System.DBNull ->
                System.Data.SqlTypes.SqlDateTime.Null
            | _ ->
                new System.Data.SqlTypes.SqlDateTime(i :?> DateTime)
            
    [<SqlFunction(IsDeterministic = false, IsPrecise = true)>]
        static member BITFUN(conn : string, cmd : string, r : int, c : int) =
            let results = new ArrayList()
            let ds = funUtil.ex conn cmd
            let i = ds.Tables.[0].Rows.[r].Item(c)
            match i with
            | :? System.DBNull ->
                System.Data.SqlTypes.SqlBoolean.Null
            | _ ->
                new System.Data.SqlTypes.SqlBoolean(i :?> bool)

